ws/src/main/scala/de/thm/ii/fbs/services/persistence/CourseResultService.scala

```diff
@@ -0,0 +1,66 @@
+package de.thm.ii.fbs.services.persistence
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import de.thm.ii.fbs.model.{CourseResult, TaskResult}
+import de.thm.ii.fbs.util.DB
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.jdbc.core.JdbcTemplate
+import org.springframework.stereotype.Component
+
+import java.sql.ResultSet
+
+/**
+  *
+  */
+@Component
+class CourseResultService {
+  @Autowired
+  private implicit val jdbc: JdbcTemplate = null
+  @Autowired
+  private val courseRegistration: CourseRegistrationService = null
+  @Autowired
+  private val objectMapper: ObjectMapper = null
+
+  /**
+    * Get All Course Results
+    * @param cid Course id
+    * @return all Course Results
+    */
+  def getAll(cid: Int): List[CourseResult] = DB.query("""
+    |select u.user_id
+    |     ,u.prename
+    |     ,u.surname
+    |     ,u.email
+    |     ,u.username
+    |     ,u.global_role
+    |     ,u.alias
+    |     ,JSON_ARRAYAGG(JSON_OBJECT("task", JSON_OBJECT("id", t.task_id,
+    |                                                    "name", t.name,
+    |                                                    "deadline", t.deadline,
+    |                                                    "mediaType", t.media_type,
+    |                                                    "description", t.description),
+    |                                                    "attempts", coalesce(submissions.attempts, 0),
+    |                                                    "passed", coalesce(submissions.passed, 0)
+    |    )) as results
+    |     ,IF(sum(passed) = count(distinct t.task_id), 1, 0) as passed
+    |from user u
+    |         left join user_course uc using (user_id)
+    |         inner join task t using (course_id)
+    |         left join (
+    |            select uts.user_id
+    |                 ,uts.task_id
+    |                 ,count(distinct uts.submission_id) as attempts
+    |                 ,max(IF(cr.exit_code = 0, 1, 0)) as passed
+    |            from user_task_submission uts
+    |                     left join checkrunner_configuration cc using (task_id)
+    |                     left join checker_result cr using (submission_id, configuration_id)
+    |            group by uts.user_id, uts.task_id
+    |) as submissions using (user_id, task_id)
+    |where course_id = ?
+    |group by u.user_id;
+    |""".stripMargin, (res, _) => parseResult(res), cid)
+
+  private def parseResult(res: ResultSet): CourseResult = CourseResult(
+    courseRegistration.parseUserResult(res), res.getBoolean("passed"), objectMapper.readValue(res.getString("results"), classOf[Array[TaskResult]]).toList
```

Comment:
I had to change `List` to `Array` because otherwise the following error would occur when using the object in Scala (e.g. in `/api/v1/courses/:id/evaluation/results`): 

```
java.lang.ClassCastException: class scala.collection.immutable.HashMap cannot be cast to class de.thm.ii.fbs.model.Task (scala.collection.immutable.HashMap and de.thm.ii.fbs.model.Task are in unnamed module of loader 'app')


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.parseTasksResult(EvaluationContainerService.scala:137) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.parseResult(EvaluationContainerService.scala:127) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.$anonfun$getOne$1(EvaluationContainerService.scala:52) ~[ws-1.0-SNAPSHOT.jar:na]


	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:94) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:61) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:679) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:617) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:669) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:700) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:712) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:768) ~[spring-jdbc-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at de.thm.ii.fbs.util.DB$.query(DB.scala:32) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.getOne(EvaluationContainerService.scala:46) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.$anonfun$createContainer$2(EvaluationContainerService.scala:93) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.$anonfun$createContainer$2$adapted(EvaluationContainerService.scala:93) ~[ws-1.0-SNAPSHOT.jar:na]


	at scala.Option.flatMap(Option.scala:283) ~[scala-library-2.13.1.jar:na]


	at de.thm.ii.fbs.services.persistence.EvaluationContainerService.createContainer(EvaluationContainerService.scala:93) ~[ws-1.0-SNAPSHOT.jar:na]


	at de.thm.ii.fbs.controller.CourseEvaluationController.createContainer(CourseEvaluationController.scala:103) ~[ws-1.0-SNAPSHOT.jar:na]


	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]


	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]


	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]


	at java.base/java.lang.reflect.Method.invoke(Method.java:564) ~[na:na]


	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]


	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE] at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.jasig.cas.client.util.AssertionThreadLocalFilter.doFilter(AssertionThreadLocalFilter.java:50) ~[cas-client-core-3.5.1.jar:3.5.1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.jasig.cas.client.util.HttpServletRequestWrapperFilter.doFilter(HttpServletRequestWrapperFilter.java:71) ~[cas-client-core-3.5.1.jar:3.5.1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:367) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1639) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[na:na]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at java.base/java.lang.Thread.run(Thread.java:832) ~[na:na]
```

If you know a better solution for this problem, you are welcome to adapt it :smile: